src/pkg/syscall/syscall_akaros.go

```diff
@@ -272,15 +272,27 @@ func Exit(exitcode int) {
 	proc_destroy(int(parlib.Procinfo.Pid), exitcode)
 }
 
-//sys pipe(p *[2]_C_int, flags int) (err error)
 func Pipe(p []int, flags int) (err error) {
 	if len(p) != 2 {
 		return EINVAL
 	}
-	var pp [2]_C_int
-	err = pipe(&pp, flags)
-	p[0] = int(pp[0])
-	p[1] = int(pp[1])
+	dirfd, err := Open("#pipe", O_PATH | (flags & O_CLOEXEC), 0 )
+	if err != nil {
+		return
+	}
+	dfd, err := Openat(dirfd, "data", O_RDWR | flags, 0 )
+	if err != nil {
+		Close(dirfd);
+		return
+	}
+	d1fd, err := Openat(dirfd, "data1", O_RDWR | flags, 0 )
+	if err != nil {
+		Close(dfd);
+		Close(dirfd);
+		return
+	}
+	p[0] = dfd
+	p[1] = d1fd
 	return
```

Comment:
You need to Close(dirfd) before you return.

Basically, you opened dirfd, from which you were able to open dfd and d1fd.  At that point, you don't need dirfd anymore.  That's what the glibc version does too.

Also, if you want, you can 'defer' the closing of dirfd, so you don't have to do that one in the error cases either.